File: server/middleware/errorHandler.js
const errorCodes = {
  400: 'BAD_REQUEST',
  401: 'UNAUTHORIZED',
  404: 'NOT_FOUND',
  409: 'CONFLICT',
  422: 'VALIDATION_ERROR',
  500: 'INTERNAL_SERVER_ERROR'
};

class AppError extends Error {
  constructor(message, status = 500, code = null) {
    super(message);
    this.status = status;
    this.code = code || errorCodes[status] || 'INTERNAL_SERVER_ERROR';
  }
}

function errorHandler(err, req, res, next) {
  if (res.headersSent) {
    return next(err);
  }

  const status = err.status || 500;
  const message = status === 500 && process.env.NODE_ENV === 'production'
    ? 'Internal server error'
    : err.message || 'Internal server error';

  if (status >= 500) {
    console.error(err);
  }

  return res.status(status).json({
    error: true,
    message,
    code: err.code || errorCodes[status] || 'INTERNAL_SERVER_ERROR'
  });
}

module.exports = { AppError, errorHandler };

File: server/middleware/error_handler.go
package middleware

import (
	"encoding/json"
	"errors"
	"net/http"
)

var errorCodes = map[int]string{
	http.StatusBadRequest:          "BAD_REQUEST",
	http.StatusUnauthorized:        "UNAUTHORIZED",
	http.StatusNotFound:            "NOT_FOUND",
	http.StatusConflict:            "CONFLICT",
	http.StatusRequestEntityTooLarge: "PAYLOAD_TOO_LARGE",
	http.StatusUnprocessableEntity: "VALIDATION_ERROR",
	http.StatusInternalServerError: "INTERNAL_SERVER_ERROR",
}

// Meta contains response metadata shared by list endpoints.
type Meta struct {
	Total int `json:"total"`
}

// APIResponse is the success response envelope used by all handlers.
type APIResponse struct {
	Success bool `json:"success"`
	Data    any  `json:"data"`
	Meta    any  `json:"meta,omitempty"`
}

// APIError is the error response envelope used by all handlers.
type APIError struct {
	Error   bool   `json:"error"`
	Message string `json:"message"`
	Code    string `json:"code"`
}

// AppError carries an HTTP status and stable API error code.
type AppError struct {
	Message string
	Status  int
	Code    string
}

// Error returns the API-safe error message.
func (e AppError) Error() string {
	return e.Message
}

// NewAppError creates an application error with a stable code.
func NewAppError(message string, status int, code string) AppError {
	if code == "" {
		code = errorCodes[status]
	}
	if code == "" {
		code = "INTERNAL_SERVER_ERROR"
	}
	return AppError{Message: message, Status: status, Code: code}
}

// Respond writes a JSON success response envelope.
func Respond(w http.ResponseWriter, status int, data any, meta any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(APIResponse{Success: true, Data: data, Meta: meta})
}

// RespondError writes a JSON error response envelope.
func RespondError(w http.ResponseWriter, status int, message string, code string) {
	if code == "" {
		code = errorCodes[status]
	}
	if code == "" {
		code = "INTERNAL_SERVER_ERROR"
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(APIError{Error: true, Message: message, Code: code})
}

// HandleError maps an error to the API error envelope.
func HandleError(w http.ResponseWriter, err error) {
	var appErr AppError
	if errors.As(err, &appErr) {
		RespondError(w, appErr.Status, appErr.Message, appErr.Code)
		return
	}
	RespondError(w, http.StatusInternalServerError, "Internal server error", "INTERNAL_SERVER_ERROR")
}
